types/validator_set.go

```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
 		panic("Cannot call IncrementProposerPriority with non-positive times")
 	}
 
-	const shiftEveryNthIter = 10
+	// cap the difference between priorities to be proportional to 2*totalPower by
+	// re-normalizing priorities, i.e., rescale all priorities by multiplying with:
+	//  2*totalVotingPower/(maxPriority - minPriority)
+	diff := computeMaxMinPriorityDiff(vals)
+	threshold := 2 * vals.TotalVotingPower()
+	// the 2nd check (threshold > 0) is merely a sanity check which could be
+	// removed if all tests would init. voting power appropriately;
+	// i.e. threshold should always be > 0
+	if diff > threshold && threshold > 0 {
+		// div = Ceil((maxPriority - minPriority) / 2*totalVotingPower)
+		// threshold > 0 and diff > threshold guarantees (diff / threshold > 0):
+		div := int64(math.Ceil(float64(diff) / float64(threshold)))
+		vals.dividePrioritiesBy(div)
```

Comment:
we could probably inline `dividePrioritiesBy` here as it's just a for-loop and we would save one casting to int and back to float64 inside of `dividePrioritiesBy`.